alibi_detect/saving/loading.py

```diff
@@ -68,7 +62,7 @@
 ]
 
 
-def load_detector(filepath: Union[str, os.PathLike], **kwargs) -> Detector:
+def load_detector(filepath: Union[str, os.PathLike], **kwargs) -> ConfigDetector:
```

Comment:
I'm assuming the name change from `Detector` to `ConfigDetector` is to signify that detectors that have config saving implemented are returned? But would this be misleading since legacy detectors are still returned from the `load_detector` function call? Should we just stick with `Detector` or is there something I'm missing?

Reply:
Yeh admittedly `ConfigDetector` is a bit of a hack (hope you don't mind me saying @mauicv!). This was necessary because I implemented the `get_config` and `from_config` methods via `DriftConfigMixin` instead of as abstract methods of `BaseDetector`. The reason for this was because outlier and adversarial detectors do not yet support config save/load. 

IMO the most principled solution would either be to create a separate `BaseDriftDetector`, and define `get_config` and `from_config` in there, or alternatively, define the methods in `BaseDetector` and have the methods raise `NotImplementedError` in outlier and adversarial detectors. The `DriftConfigMixin` was intended to be a temporary measure whilst we decided what exactly the revamped outlier detectors would look like...

Reply:
Yeah, I agree w.r.t. `DriftConfigMixin` feeling hacky. It's frustrating that [intersection types](https://github.com/python/typing/issues/213) aren't implemented but it seems there quite tricky.

If we decide to stick with it however we can also use [`TypeGuards`](https://mypy.readthedocs.io/en/latest/type_narrowing.html#user-defined-type-guards). So we define a function:

```py
def _is_config_detector(detector: Union[BaseDetector, ConfigDetector]) -> TypeGuard[ConfigDetector]:
    return hasattr(detector, 'get_config')
```

That narrows the detector type. This is simpler than changing the inheritance structure of all the detectors but perhaps still less principled.

Perhaps the method of making the above more "Proper" is to use [protocols](https://docs.python.org/3/library/typing.html#typing.Protocol). See also [this](https://stackoverflow.com/a/70816342) and [this](https://stackoverflow.com/a/50320982). I haven't tried this yet though.

Reply:
I like the protocol approach and I think it is also the modern/recommended way of doing this sort of thing. As @ascillitoe mentioned, would need to check if `Protocol` and `runtime_checkable` are available in `typing_extensions` for Python 3.7 support.